Add data projections

azywot opened the pull request:
Resolves #125 & #134, the projections have been incorporated to the dashboard.
I'm pasting a ss below, highly encourage you to test it as well 🙏

![obraz](https://github.com/zuzg/drug-screening/assets/82370491/aa55e62c-2c3a-4ca9-a6f4-326082813940)

*Note: I mocked the data for testing purposes, the files can be found below and are of the same structure as the screening output.*
[experiment1.csv](https://github.com/zuzg/drug-screening/files/12905808/experiment1.csv)
[experiment2.csv](https://github.com/zuzg/drug-screening/files/12905810/experiment2.csv)
[experiment3.csv](https://github.com/zuzg/drug-screening/files/12905811/experiment3.csv)

zuzg commented on dashboard/data/combine.py:
```diff
@@ -6,48 +6,48 @@
-# NOTE: to be removed
-def combine_assays(
-    dataframes: list[pd.DataFrame],
+def combine_assays_for_projections(
```
duplicate method (it exists also in `preprocess.py`)

azywot replied:
Good point, sorry for that!

zuzg commented on dashboard/data/preprocess.py:
```diff
@@ -17,44 +19,76 @@ def transform(self, X: np.ndarray) -> np.ndarray:
+        processed_dfs = []
+        GROUP_BY_COLUMNS = [
+            "EOS",
+            "Source Plate Barcode",
+            "Source Well",
+            "Destination Plate Barcode",
+            "Destination Well",
```
do we need source and destination? tried to use it with our initial data (from sem V) and there were just `Plate `and `Well` columns

azywot replied:
Yeah, though in the output file we have everything, hence the names.
Great that you pointed this out tho, I think it's wise to remove `GROUP_BY_COLUMNS` and group it using solely `EOS`, as it's an experiment from one screening.

zuzg commented on dashboard/pages/data_projection/page.py:
```diff
@@ -0,0 +1,25 @@
+pb = ProcessPageBuilder(name=NAME)
+STAGE_NAMES = [
+    "Files Input",
+    "Projections Visualization",
```
I would name it just Visualization, since the stage name already consists  "projection"

azywot replied:
Good point, done 👍

zuzg commented on dashboard/pages/data_projection/stages/s2_projection_display.py:
```diff
@@ -0,0 +1,76 @@
+                html.Div(
+                    children=[
+                        html.Div(
+                            id="projection-table",
```
As far as I remember, the table automatically adjusted for the compounds currently visible on the plot. It doesn't work now, not sure if it was intended

azywot replied:
There are couple of things tbd in order to match the previous version:
- dynamic data table connected with plot 
- add control values 
- loading sign (and this one to be set everywhere imo)
- ACTIVATION/INHIBITION choice for every experiment

Will add it to the issues & can take care of it, just wanted to provide the basic version; I can turn it into a draft and work on it further too.

zuzg replied:
ok sure, we may work on that later

zuzg commented on dashboard/pages/data_projection/stages/s3_report.py:
```diff
@@ -0,0 +1,19 @@
+            className="d-flex justify-content-between",
+            children=[
+                html.Button(
+                    "Save projection data as CSV",
```
imo centered button would look better

azywot replied:
I followed the convention we have now 😬

zuzg reviewed (commented):
Looks good, added smol comments

Tremirre reviewed (approved):
LGTM, GJ 👌

azywot commented:
Added the responsive table and wrapped the elements into loadings, now it looks like this:

![obraz](https://github.com/zuzg/drug-screening/assets/82370491/1746add6-ae36-4d24-aac1-2533452b6683)

⚠️minor issue but still an issue⚠️
One challenge I'm facing rn is to adjust the loading sign to be in line with the figure's one, tried some styling or mocking an empty table, but didn't come up with anything sensible, as once the table has < 10 records, the loading moves up too:
![obraz](https://github.com/zuzg/drug-screening/assets/82370491/7f6b3411-102f-4722-9eea-0b23e1b89791)

azywot commented:
Update - now it includes control values and additional info about PCA:

![obraz](https://github.com/zuzg/drug-screening/assets/82370491/38def7d0-70f2-4c1c-9080-f6b7d7f03517)
![obraz](https://github.com/zuzg/drug-screening/assets/82370491/febb34f7-9673-4351-b677-137000da2227)

Tremirre commented:
I struggle with setting up the environment for this branch, as the list of the dependencies seems to be incomplete or not specific enough:

> ❯ python -m dashboard
> Traceback (most recent call last):
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\__main__.py", line 1, in <module>
>     from .app import app
>   File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\app.py", line 19, in <module>   
>     app = Dash(
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\dash.py", line 533, in __init__
>     self.init_app()
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\dash.py", line 611, in init_app
>     self.enable_pages()
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\dash.py", line 2082, in enable_pages
>     _import_layouts_from_pages(self.config.pages_folder)
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\_pages.py", line 435, in _import_layouts_from_pages        
>     spec.loader.exec_module(page_module)
>   File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\pages\data_projection\page.py", line 5, in <module>
>     from dashboard.pages.data_projection.callbacks import register_callbacks
>   File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\pages\data_projection\callbacks.py", line 11, in <module>
>     from umap import UMAP
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\umap\__init__.py", line 7, in <module>
>     from .parametric_umap import ParametricUMAP
>   File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\umap\parametric_umap.py", line 28, in <module>
>     TF_MAJOR_VERSION = int(tf.__version__.split(".")[0])
> AttributeError: module 'tensorflow' has no attribute '__version__'

I reinstalled the environment from scratch but got the same result, which implies that our _staging_ and _prod_ environments will also fail

perhaps you could add the specific versions of the dependencies on your environment and freeze them to _requirements.txt_?

azywot commented:
> I struggle with setting up the environment for this branch, as the list of the dependencies seems to be incomplete or not specific enough:
> 
> > ❯ python -m dashboard
> > Traceback (most recent call last):
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\runpy.py", line 196, in _run_module_as_main
> > return _run_code(code, main_globals, None,
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\runpy.py", line 86, in _run_code
> > exec(code, run_globals)
> > File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard__main__.py", line 1, in 
> > from .app import app
> > File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\app.py", line 19, in 
> > app = Dash(
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\dash.py", line 533, in **init**
> > self.init_app()
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\dash.py", line 611, in init_app
> > self.enable_pages()
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash\dash.py", line 2082, in enable_pages
> > _import_layouts_from_pages(self.config.pages_folder)
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\dash_pages.py", line 435, in _import_layouts_from_pages
> > spec.loader.exec_module(page_module)
> > File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\pages\data_projection\page.py", line 5, in 
> > from dashboard.pages.data_projection.callbacks import register_callbacks
> > File "C:\Users\sbart\OneDrive - put.poznan.pl\Semestr V\Problem Classes\drug-screening\dashboard\pages\data_projection\callbacks.py", line 11, in 
> > from umap import UMAP
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\umap__init__.py", line 7, in 
> > from .parametric_umap import ParametricUMAP
> > File "D:\Programs\Anaconda\envs\drug-screening\lib\site-packages\umap\parametric_umap.py", line 28, in 
> > TF_MAJOR_VERSION = int(tf.**version**.split(".")[0])
> > AttributeError: module 'tensorflow' has no attribute '**version**'
> 
> I reinstalled the environment from scratch but got the same result, which implies that our _staging_ and _prod_ environments will also fail
> 
> perhaps you could add the specific versions of the dependencies on your environment and freeze them to _requirements.txt_?

Added the version of `umap-learn` package I have locally

Tremirre commented:
> > [...] I reinstalled the environment from scratch but got the same result, which implies that our _staging_ and _prod_ environments will also fail
> > perhaps you could add the specific versions of the dependencies on your environment and freeze them to _requirements.txt_?
> 
> Added the version of `umap-learn` package I have locally

I think you'd have to input the version of `tensorflow` installed - umap does not explicitly install it, but its an extra requirement for parametric umap that we use from this package

Tremirre commented on dashboard/pages/data_projection/callbacks.py:
```diff
@@ -0,0 +1,306 @@
+    last_modified: int,
+    stored_uuid: str,
+    file_storage: FileStorage,
+) -> html.Div:
```
the initial callback for each process has to set `stored_uuid` if it does not exists (that happens whenever the user simply does not have the `stored_uuid` set but wishes to use the process), because without it the processing will fail

As such you have to return the `stored_uuid` as well from this callback and save it to appropriate uuid holding `State`

azywot replied:
Ahh, okay I see, thanks for the explanation, it's corrected now 👍

Tremirre commented on dashboard/pages/data_projection/callbacks.py:
```diff
@@ -0,0 +1,306 @@
+    ).set_columns_for_projection(projection_columns)
+
+    for projector, name in PROJECTION_SETUP:
+        assays_preprocessor.apply_projection(projector, name)
```
doesn't this apply the same projection twice?

azywot replied:
One's for `fit_transform`, the other - solely `transform` for projecting control values.
On the other hand, I can just modify `apply_projection` to make it in a single step - will do it that way to make it more intuitive

Tremirre commented on requirements.txt:
```diff
@@ -8,7 +8,7 @@ scikit-learn
 jenkspy
 numpy
-umap-learn
+umap-learn==0.5.3
```
missing tensorflow dependency

Tremirre reviewed (commented):
added some comments + pointed out one bug

btw imo we can remove __LEGACY folder at this point

azywot commented:
> added some comments + pointed out one bug
> 
> btw imo we can remove __LEGACY folder at this point

Thanks for the comments, corrected everything 🙏
